### PR TITLE
Automatically update initramfs when regression tests are changed

### PR DESCRIPTION
**Pull request by lrh2000**
Example:
 - If no regression tests are changed, initramfs will not be regenerated:
```
root@localhost:~/asterinas# make -s initramfs
root@localhost:~/asterinas# 
```
 - If regression tests are changed, tests will be rebuilt and initramfs will be regenerated:
```
root@localhost:~/asterinas# touch regression/apps/network/source.c 
root@localhost:~/asterinas# make -s initramfs
CC <= /root/asterinas/regression/apps/network/../../build/initramfs/regression/network/source
Generating the initramfs image...
40377 blocks
root@localhost:~/asterinas# 
```
 - Everything works properly even if regression tests include some header files:
```
root@localhost:~/asterinas# cat regression/apps/network/source.c 
#include "header.h"

int main(void) { return 0; }
root@localhost:~/asterinas# make -s initramfs
root@localhost:~/asterinas# touch regression/apps/network/header.h 
root@localhost:~/asterinas# make -s initramfs
CC <= /root/asterinas/regression/apps/network/../../build/initramfs/regression/network/source
Generating the initramfs image...
40377 blocks
root@localhost:~/asterinas# 
```
 - If regression scripts are changed, they will be copied and initramfs will be regenerated:
```
root@localhost:~/asterinas# touch regression/apps/scripts/network.sh 
root@localhost:~/asterinas# make -s initramfs
Generating the initramfs image...
40377 blocks
root@localhost:~/asterinas# 
```

Also, the following changes have been made to all `Makefile`s:
 - The `.PHONY` should be next to the actual target instead of at the beginning. I think this makes things clearer.
 - Directories should not be a normal requirement. Since `make` checks the modification time of directories instead of the files under them, the target will only be rebuilt if the directories themselves change (e.g. some new files are added or deleted), which is not normally expected behavior. Instead, make directories order-only prerequisites, so that they are called only once, usually the first time the directory needs to be created. See [the example in the official documentation](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html).
 - Use rules when we can, instead of relying on loops in shell scripts. See the [official documentation](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html) for the reasons.

### CODE REVIEW

**Comment by liqinggd** [comment]
1. I think the following commits [Fix dependencies of regression scripts](https://github.com/asterinas/asterinas/pull/659/commits/464f972f429cd4b50409375daa5c455af6aa4125),  [Fix dependencies of regression tests](https://github.com/asterinas/asterinas/pull/659/commits/3f717054ad212752e17fe192146ae7c897abe428),
[List regression modules as targets](https://github.com/asterinas/asterinas/pull/659/commits/2661d912eb29527892d6a4cbc0ef7a14b31bc29f),
[Add missing spaces in EXTRA_C_FLAGS](https://github.com/asterinas/asterinas/pull/659/commits/7dd30845eb781fd8a2e1ec60561420bbf945a7a9) can be squashed into just one commit, because these commits are all about the refining compiling of regression test apps and scripts.

2.

File: regression/Makefile [hunk 1]
```diff
@@ -82,18 +82,27 @@ $(INITRAMFS)/benchmark:
...
+.PHONY: $(INITRAMFS)/opt/syscall_test
 $(INITRAMFS)/opt/syscall_test:
 	@make --no-print-directory -C syscall_test
 
+.PHONY: $(INITRAMFS_IMAGE)
+$(INITRAMFS_IMAGE): initramfs_filelist
+	@if ! cmp -s $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).old ; then \
+		echo "Generating the initramfs image..."; \
+		cp $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).old; \
+		(cd $(INITRAMFS); find . | cpio -o -H newc | gzip) > $@; \
```
Comment (liqinggd) [inline regression/Makefile:94]: Why not just cat the file `initramfs_filelist` and call cpio to generate the image?

File: regression/Makefile [hunk 2]
```diff
@@ -82,18 +82,27 @@ $(INITRAMFS)/benchmark:
...
+	fi
+
+.PHONY: initramfs_filelist
```
Comment (liqinggd) [inline regression/Makefile:97]: I think the name `initramfs_filelist` is redundant, You can just use `$(INITRAMFS_FILELIST)`

File: regression/Makefile [hunk 3]
```diff
@@ -82,18 +82,27 @@ $(INITRAMFS)/benchmark:
...
 
+.PHONY: $(INITRAMFS)/opt/syscall_test
 $(INITRAMFS)/opt/syscall_test:
 	@make --no-print-directory -C syscall_test
 
+.PHONY: $(INITRAMFS_IMAGE)
+$(INITRAMFS_IMAGE): initramfs_filelist
+	@if ! cmp -s $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).old ; then \
+		echo "Generating the initramfs image..."; \
+		cp $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).old; \
```
Comment (liqinggd) [inline regression/Makefile:93]: cp -f cp $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).previous

The `previous` is a more accurate expression.

File: regression/Makefile (hunk 1 above)
Comment (lrh2000) [reply regression/Makefile:94]: `$(INITRAMFS_FILELIST)` contains the modification time to track when these files are updated in initramfs. The modification time must be removed before passing its contents to `cpio`.

I changed it to `cut -d "" -f 2-` ("2-" to allow spaces in filenames, although we should avoid spaces there), but I'm not sure if that's better or not. Both are acceptable to me.

Example:
```
> cat regression/build/initramfs.filelist | head -10
1709231337.5710777910 .
1709231334.8777257870 ./etc
1709231334.8777257870 ./etc/passwd
1709231334.8777257870 ./etc/group
1709231336.4677368110 ./lib
1709231336.4944036620 ./lib/x86_64-linux-gnu
1709231336.4744035230 ./lib/x86_64-linux-gnu/libc.so.6
1709231336.4777368790 ./lib/x86_64-linux-gnu/libstdc++.so.6
1709231336.4810702360 ./lib/x86_64-linux-gnu/libm.so.6
1709231336.4810702360 ./lib/x86_64-linux-gnu/libgcc_s.so.1
```
```
> cat regression/build/initramfs.filelist | cut -d " " -f 2- | head -10
.
./etc
./etc/passwd
./etc/group
./lib
./lib/x86_64-linux-gnu
./lib/x86_64-linux-gnu/libc.so.6
./lib/x86_64-linux-gnu/libstdc++.so.6
./lib/x86_64-linux-gnu/libm.so.6
```

**Comment by lrh2000** [comment]
> I think the following commits [Fix dependencies of regression scripts](https://github.com/asterinas/asterinas/pull/659/commits/464f972f429cd4b50409375daa5c455af6aa4125), [Fix dependencies of regression tests](https://github.com/asterinas/asterinas/pull/659/commits/3f717054ad212752e17fe192146ae7c897abe428), [List regression modules as targets](https://github.com/asterinas/asterinas/pull/659/commits/2661d912eb29527892d6a4cbc0ef7a14b31bc29f), [Add missing spaces in EXTRA_C_FLAGS](https://github.com/asterinas/asterinas/pull/659/commits/7dd30845eb781fd8a2e1ec60561420bbf945a7a9) can be squashed into just one commit, because these commits are all about the refining compiling of regression test apps and scripts.

Thanks for your comments. I have no objection to squashing these commits. While the resulting commit may seem less atomic, given the small number of lines changed, it does *not* make understanding git history much harder.

File: regression/Makefile [hunk 4]
```diff
@@ -82,18 +83,31 @@ $(INITRAMFS)/benchmark:
...
+$(SYSCALL_TEST_DIR):
 	@make --no-print-directory -C syscall_test
 
+.PHONY: $(INITRAMFS_IMAGE)
+$(INITRAMFS_IMAGE): $(INITRAMFS_FILELIST)
+	@if ! cmp -s $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).previous ; then \
+		echo "Generating the initramfs image..."; \
+		cp -f $(INITRAMFS_FILELIST) $(INITRAMFS_FILELIST).previous; \
+		( \
+			cd $(INITRAMFS); \
```
Comment (liqinggd) [inline regression/Makefile:96]: I think `cd $(INITRAMFS);` is not necessary here.

File: regression/Makefile [hunk 5]
```diff
@@ -82,18 +83,31 @@ $(INITRAMFS)/benchmark:
...
+			cut -d " " -f 2- $(INITRAMFS_FILELIST) | \
```
Comment (liqinggd) [inline regression/Makefile:97]: Add some comments here to explain that we need to exclude the timestamp.

Comment (lrh2000) [reply regression/Makefile:101]: ~~How can I add comments when the trailing `\` always extends the comments to the next line?~~

~~I've been trying to do this, but haven't found a way yet.~~

This is a bug in Vim's highlighting of comments. Comments can be added here without problems.

File: regression/Makefile (hunk 4 above)
Comment (lrh2000) [reply regression/Makefile:96]: We need to give `cpio` relative paths, not just the path on the host file system, but the path inside the generated initramfs.

So if we omit `cd $(INITRAM)`, `cpio` has no way to locate the relative path.

```
Generating the initramfs image...
cpio: ./lib: Cannot stat: No such file or directory
cpio: ./lib/x86_64-linux-gnu: Cannot stat: No such file or directory
cpio: ./lib/x86_64-linux-gnu/libc.so.6: Cannot stat: No such file or directory
```

Comment (liqinggd) [reply regression/Makefile:96]: OK, thanks for the explanation

**Comment by liqinggd** [comment]
LGTM

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**